### PR TITLE
Add kafka queue

### PR DESCRIPTION
**Pull request by maikelmclauflin**
- [ ] added kafka queue consumption for referrals and settlements

### CODE REVIEW

File: bat-utils/lib/runtime-kafka.js [hunk 1]
```diff
@@ -21,29 +21,106 @@ class Kafka {
...
+    //   analytics,
+    //   health
+    // )
+  }
+
+  async producer () {
+    if (this._producer) {
+      return this._producer
+    }
+    await this.connect()
```
Comment (keur) [inline bat-utils/lib/runtime-kafka.js:55]: there should probably be some assertion that `this._producer` is set after this call. probably i'd format it like this

```
if (!this._producer) {
  await this.connect()
  assert(this._producer)
}
return this_.producer
```

File: bin/migrate-mongo-to-kafka.js [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+      referralCode = '',
+      owner,
+      publisher,
+      transactionId,
+      groupId,
+      platform
+    }) => ({
+      downloadId,
+      downloadTimestamp: new Date((downloadTimestamp || finalized).toISOString()),
+      finalizedTimestamp: new Date(finalized.toISOString()),
```
Comment (keur) [inline bin/migrate-mongo-to-kafka.js:64]: should both of these `toISOString()` calls be on the date object?

File: bin/migrate-mongo-to-kafka.js [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+      return transaction.id.settlement(settlementId, normalizedChannel, type)
+    })
+    const { rows } = await runtime.postgres.query(`
+    select count(*)
+    from transactions
+    where id = any($1::uuid[])`, [ids], true)
+    console.log('count', rows[0].count)
+    if (!(+rows[0].count)) {
+      return []
+    }
```
Comment (keur) [inline bin/migrate-mongo-to-kafka.js:92]: You are doing the same query twice. De-dup in it's own function?

```
async getTransactionsFromIds(txids) {
    const { rows } = await runtime.postgres.query(`
    select count(*)
    from transactions
    where id = any($1::uuid[])`, [txids], true)
    console.log('count', rows[0].count)
    if (!(+rows[0].count)) {
      return []
    }
}

File: bin/migrate-mongo-to-kafka.js [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+      amount: amount.toString(),
+      probi: probi.toString(),
+      fees: fees.toString(),
+      fee: fee.toString(),
+      commission: commission.toString()
+    }))
+  })
+}
+
+async function connectToKafka (collectionName, key, coder, transform) {
```
Comment (keur) [inline bin/migrate-mongo-to-kafka.js:125]: nit: transformFn

File: bin/migrate-mongo-to-kafka.js [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  // filter out the empty strings
+  const distinct = ids.filter((item) => item)
+  for (let i = 0; i < distinct.length; i += 1) {
+    const targetId = distinct[i]
+    const documents = await collection.find({
+      [key]: targetId,
+      migrated: { $ne: true }
+    })
+    if (!documents.length) {
+      console.log('no documents found', collectionName)
```
Comment (keur) [inline bin/migrate-mongo-to-kafka.js:137]: change to `console.log('no documents found for id: ' + targetId)`

File: bin/migrate-mongo-to-kafka.js [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+      continue
+    }
+    const messages = await transform(documents)
+    if (!messages.length) {
+      console.log('no messages transformed', collectionName)
```
Comment (keur) [inline bin/migrate-mongo-to-kafka.js:142]: change to console.log('no documents transformed for id: ' + targetId)

File: eyeshade/lib/queries.js [hunk 7]
```diff
@@ -87,3 +88,16 @@ WHERE
     geo_referral_groups.active_at <= $1
 AND countries.group_id = geo_referral_groups.id;`
 }
+
+function getActiveCountryGroups () {
```
Comment (keur) [inline eyeshade/lib/queries.js:92]: weird formatting where some of the postgres keywords are uppercase, some lowercase.

**Review by keur — COMMENTED** [review]
logic lgtm for the referral stuff. just some minor comments.

File: eyeshade/lib/queries.js (hunk 7 above)
Comment (maikelmclauflin) [reply eyeshade/lib/queries.js:92]: fixed in: eec22ca

File: bin/migrate-mongo-to-kafka.js (hunk 5 above)
Comment (maikelmclauflin) [reply bin/migrate-mongo-to-kafka.js]: fixed in: eec22ca

File: bin/migrate-mongo-to-kafka.js (hunk 4 above)
Comment (maikelmclauflin) [reply bin/migrate-mongo-to-kafka.js]: fixed in: eec22ca

File: bin/migrate-mongo-to-kafka.js (hunk 3 above)
Comment (maikelmclauflin) [reply bin/migrate-mongo-to-kafka.js:86]: fixed in: eec22ca

File: bin/migrate-mongo-to-kafka.js (hunk 2 above)
Comment (maikelmclauflin) [reply bin/migrate-mongo-to-kafka.js]: fixed in: eec22ca

File: bat-utils/lib/runtime-kafka.js (hunk 1 above)
Comment (maikelmclauflin) [reply bat-utils/lib/runtime-kafka.js]: i don't think that this is something we have to worry about. creating a producer is happening sync
fixed in: eec22ca

**Review by keur — APPROVED** [review]
LGTM once you get CI fixed up.

File: bat-utils/lib/runtime-kafka.js [hunk 8]
```diff
@@ -21,29 +21,106 @@ class Kafka {
...
+    producer.on('error', error => {
       console.error(error)
       if (this.runtime.captureException) {
         this.runtime.captureException(error)
       }
     })
-    await this.producer.connect()
+    await producer.connect()
+    // DEBUGGING
+    // const analytics = await producer.enableAnalytics()
```
Comment (husobee) [inline bat-utils/lib/runtime-kafka.js:40]: Clean up these comments?

**Review by husobee — DISMISSED** [review]
lgtm, just one note about commented out lines.

File: bat-utils/lib/runtime-kafka.js (hunk 8 above)
Comment (maikelmclauflin) [reply bat-utils/lib/runtime-kafka.js]: fixed in: 315079e

**Review by husobee — APPROVED** [review]

**Pull request merged**